### PR TITLE
Exercise 12

### PR DESCRIPTION
**Pull request by egrfly**

### CODE REVIEW

File: .github/workflows/ci-cd-pipeline.yml [hunk 1]
```diff
@@ -70,9 +70,38 @@ jobs:
...
+      TF_VAR_resource_group_name: ${{ secrets.TF_VAR_RESOURCE_GROUP_NAME }}
+      TF_VAR_secret_key: ${{ secrets.TF_VAR_SECRET_KEY }}
+      TF_VAR_webapp_name: ${{ secrets.TF_VAR_WEBAPP_NAME }}
     steps:
-      - name: Make post request to Azure app webhook URL
-        run: curl -f -X POST '${{ secrets.AZURE_APP_WEBHOOK_URL }}'
+      - name: Checkout
+        uses: actions/checkout@v3
+      - name: Log into Azure using service principal
+        run: az login --service-principal --username $ARM_CLIENT_ID --password $ARM_CLIENT_SECRET --tenant $ARM_TENANT_ID
```
Comment (JackMead) [inline .github/workflows/ci-cd-pipeline.yml:96]: You should find this step isn't _strictly_ necessary, as long as you have these values labelled correctly (and I believe these environment variables are all so) then Terraform will silently pull them from the environment anyway, but no harm in doing so!

File: terraform/variables.tf [hunk 2]
```diff
@@ -0,0 +1,39 @@
+variable "prefix" {
+  description = "The prefix used for all resources in this environment"
+  default     = "prod"
```
Comment (JackMead) [inline terraform/variables.tf:3]: Very minor, but could be worth considering specifying the `type` for each of these variables explicitly - when not specified any type can be passed through, but given that all of these are strings I suspect there's not much space for confusion as it is

File: .github/workflows/ci-cd-pipeline.yml [hunk 3]
```diff
@@ -70,9 +70,38 @@ jobs:
...
+      - name: Terraform init
+        working-directory: terraform
+        run: terraform init -backend-config=<(python -m generate_backend_config)
```
Comment (JackMead) [inline .github/workflows/ci-cd-pipeline.yml:99]: Generating the config this way is nice, but _potentially_ overkill while we're just using a single environment - even if we wanted to have multiple, would this actually be more helpful than just having a folder with several backend-configs in, and we point to the correct one?

It does let us keep some of the values secret, but I'd argue they aren't sensitive values

```

**Review by JackMeadDev — APPROVED** [review]
Looking good; I can see the pipeline's reporting Terraform completing happily, and the TF itself looks clean and to the point 😄

**Pull request merged**